### DISCUSSION TITLE
How can SAHI do faster inference?

### DISCUSSION THREAD

**Discussion by rusvagzur** [obss/sahi#972] [Ideas]
Hi @fcakyon ,

I am currently using SAHI with YOLOv8. I am not completely sure that SAHI runs batch inference, since I have observed in the code that only num_batch=1 is supported and that inferences can take up to seconds per complete image depending on the number of slices.

I was wondering how could this package be improved in order to run batch inference in the GPU. Probably also doing the slices in the GPU directly when the original image is loaded.

Is batch inference dependant on YOLOv8? or is SAHI creating the bottleneck? Several seconds for the prediction of a single image doesn't justify higher small object detection for some real-world/real-time applications like security applications.

Best regards

**Comment by andressrodrl** [comment]
@rusvagzur 

I´ve modified it to run inference with slice batches: https://github.com/andresinsitu/sahi_custom/tree/batch_inf
Added the argument batch to the predict and get_sliced_prediction function. It is working for me speeding it a bit.

At the moment the batch has to be a divisor of the number of slices, otherwise it leaves some slice without inference(but it works regardless)

**Reply by rusvagzur** [reply]
@andresinsitu Nice implementation. However, something is off. When I set the batch number and then the yolov8 model call with verbose=True, being batch number and number of slices the same, yolo is only called once. When I set the batch=1, then yolo is called multiple times, one per slice. So I am suspecting that when I introduce your changes, yolo is only performed in the original image and no prediction in the slices is happening.

**Reply by rusvagzur** [reply]
@andresinsitu I think the problem was in models/yolov8.py. `num_group = int(num_slices / num_batch)` should be:

`num_group = 1 if num_batch >= num_slices else int(num_slices / num_batch)`

    
 To avoid num_group=0 when the batch number is higher than the number of slices. In my case, I have num_slices=12.
 
 Once that is fixed, I get the following error when using `num_batch>=6`:
 
 ```
sahi/models/yolov8.py", line 87, in <listcomp>
    result.masks.data[result.boxes.data[:, 4] >= self.confidence_threshold],
AttributeError: 'NoneType' object has no attribute 'data'
```

And the following error when `num_batch<6`:

```
sahi/models/yolov8.py", line 101, in perform_inference
    self._original_shape = image.shape
AttributeError: 'list' object has no attribute 'shape'

```

**Reply by rusvagzur** [reply]
I have solved the first one by adding to `result.masks.data[result.boxes.data[:, 4] >= self.confidence_threshold]` the following

`result.masks.data[result.boxes.data[:, 4] >= self.confidence_threshold] if result.masks is not None else result.boxes.data[result.boxes.data[:, 4] >= self.confidence_threshold],`

The second error is solved with:

`self._original_shape = image[0].shape if len(prediction_result)>1 else image.shape` at the end of perform_inference.

Now it would be nice if the standard prediction done in sahi/predict.py, which is the prediction done to the whole downscaled image, would be done together with the batch when the batch number and the number of slices is the same. This would make SAHI as efficient as possible! This is where this happens now in predict.py:

```
# perform standard prediction
if num_slices > 1 and perform_standard_pred:
    prediction_result = get_prediction(
    image=[image],
    detection_model=detection_model,
    shift_amount=[0, 0],
    full_shape=None,
    postprocess=None,
)
object_prediction_list.extend(prediction_result.object_prediction_list)
```
But ideally should be done with something like this (also in predict.py):

```
prediction_result = get_prediction(
    image=image_list[0:num_batch]+[image], #being image the original image
```
Could you give me a hand with this @andresinsitu ?

**Reply by dceluis** [reply]
hey. I don't know how helpful this will be but I came up with a very small proof of concept using sahi + yolov8 to run batch inference. it's not a complete fork, more like a somewhat compatible replacement for get_sliced_prediction.

it probably only works for my use case but if you want to have a look and maybe grab some ideas here it is:
https://github.com/dceluis/sahi_batched

**Reply by bshakya77** [reply]
Hello @dceluis,

I observed the same issue that performing inference per slice is time-consuming. So, better if we could use batch in slices level. If possible, can I know how much improvement you get in inference time using your modification compared to the original code? Also, does your implementation affect the mAP or accuracy value?

You feedback is appreciated.

Regards,
Bijay

**Reply by dceluis** [reply]
Hi @bshakya77, It has been quite a while since I ran SAHI but I do not recall any significant accuracy drop, which was very surprising to me at the time. Even more surprising that the official SAHI library has not implemented it yet.

Regarding performance, I think the speedup was some number between full parallelism and sequential processing, with diminishing returns. So if using 16 slice batches the speedup was around 6x or 7x speedup, not 16x. I don't think my code is fully optimized, and I think I was hitting different bottlenecks like memory availability (which is good).

I would recommend doing your own testing, but for my use case, the mentioned speedup was very worthy.

Regards

**Reply by bshakya77** [reply]
Hello @dceluis,

Sure, I will take a look at it once. 

Thank you.

Regards,
Bijay

**Reply by bshakya77** [reply]
Hello @dceluis,

I am getting this error when I try to run your code. 
```
from sahi_batched.models import Yolov8DetectionModel
from sahi_batched import get_sliced_prediction_batched
```
![error](https://github.com/user-attachments/assets/3bd21108-9737-4d8c-8451-444d565eaa75)

I searched for that file in the original `sahi `repository. But, I could not find it. Could you please clarify the issue?

Thank you.

Regards,
Bijay

**Reply by dceluis** [reply]
Make sure SAHI is also installed. Current sahi versions may not be compatible, so it's possible you would have to lock to an older version of SAHI or tweak the batched library to be compatible with current versions.

**Reply by bshakya77** [reply]
@dceluis,

When I reviewed your code and ran the prediction using the default and your get_prediction_batched() method, it does reduce the inference time. But, when I printed the shift amount, it uses the default value [[0,0]]. I would like to know how it is performing the batch processing of the image.

I appreciate your feedback.

Regards,
Bijay

**Reply by dceluis** [reply]
@bshakya77 

Hi Bijay,

If I recall correctly, I’m also optimizing by skipping some PIL image conversions (e.g., using PseudoPIL), so you might still see improvements even if the image is too small to slice (as indicated by the [[0,0]] shift values).
